inst/include/dust/distr/poisson.hpp

```diff
@@ -9,107 +9,131 @@ namespace distr {
+  // that it is close to the inverse CDF F^-1(x), compute the following
+  // steps:
+  //
+  // 1) Generate U and V, two independent random variates. Set U = U - 0.5
+  // (this step isn't strictly necessary, but is done to make some
+  // calculations symmetric and convenient. Henceforth, G is defined on
+  // [-0.5, 0.5]).
+  //
+  // 2) If V <= alpha * F'(G(U)) * G'(U), return floor(G(U)), else return
+  // to step 1. alpha is the acceptance probability of the rejection
+  // algorithm.
+  //
+  // For more details on transformed rejection, see:
+  // http://citeseer.ist.psu.edu/viewdoc/citations;jsessionid=1BEB35946CC807879F55D42512E5490C?doi=10.1.1.48.3054.
+  //
+  // The dominating distribution in this case:
+  //
+  // G(u) = (2 * a / (2 - |u|) + b) * u + c
+
+  int x = 0;
+  const real_t log_rate = std::log(lambda);
+
+  // Constants used to define the dominating distribution. Names taken
+  // from Hormann's paper. Constants were chosen to define the tightest
+  // G(u) for the inverse Poisson CDF.
+  const real_t b = 0.931 + 2.53 * std::sqrt(lambda);
+  const real_t a = -0.059 + 0.02483 * b;
+
+  // This is the inverse acceptance rate. At a minimum (when rate = 10),
+  // this corresponds to ~75% acceptance. As the rate becomes larger, this
+  // approaches ~89%.
+  const real_t inv_alpha = 1.1239 + 1.1328 / (b - 3.4);
+
+  while (true) {
+    real_t u = dust::unif_rand<real_t>(rng_state);
+    u -= 0.5;
+    real_t v = dust::unif_rand<real_t>(rng_state);
+
+    real_t u_shifted = 0.5 - std::fabs(u);
+    int k = floor((2 * a / u_shifted + b) * u + lambda + 0.43);
+
+    if (k > dust::utils::integer_max()) {
+      // retry in case of overflow.
+      continue; // # nocov
+    }
+
+    // When alpha * f(G(U)) * G'(U) is close to 1, it is possible to
+    // find a rectangle (-u_r, u_r) x (0, v_r) under the curve, such
+    // that if v <= v_r and |u| <= u_r, then we can accept.
+    // Here v_r = 0.9227 - 3.6224 / (b - 2) and u_r = 0.43.
+    if (u_shifted >= 0.07 && v <= 0.9277 - 3.6224 / (b - 2)) {
+      x = k;
+      break;
+    }
+
+    if (k < 0 || (u_shifted < 0.013 && v > u_shifted)) {
+      continue;
+    }
+
+    // The expression below is equivalent to the computation of step 2)
+    // in transformed rejection (v <= alpha * F'(G(u)) * G'(u)).
+    real_t s = std::log(v * inv_alpha / (a / (u_shifted * u_shifted) + b));
+    real_t t = -lambda + k * log_rate -
+      dust::utils::lgamma(static_cast<real_t>(k + 1));
+    if (s <= t) {
+      x = k;
+      break;
+    }
+  }
+  return x;
+}
+
+// NOTE: we return a real, not an int, as with deterministic mode this
+// will not necessarily be an integer
+__nv_exec_check_disable__
+template <typename real_t>
+HOSTDEVICE real_t rpois(rng_state_t<real_t>& rng_state,
+                        typename rng_state_t<real_t>::real_t lambda) {
+  real_t x = 0;
   if (lambda == 0) {
     // do nothing, but leave this branch in to help the GPU
+#ifndef __CUDA_ARCH__
```

Comment:
Why do we have these includes? Is deterministic + GPU disallowed?
I appreciate that it would likely be inefficient, but might be confusing behaviour to remain stochastic

Reply:
Ok I see the check in the interface below, so this makes sense now